Multi-NUMA support

cynepco3hahue opened the pull request:
## Struct Changes

To have possibility to monitor NUMA nodes that involved into
the multi-NUMA alignment of the memory, I introduce additional
fields under the machine state type.

We need number of assignments, to know if we already allocated memory
for the container from this NUMA node, after it the code will check if
the node has some nodes under the `Nodes`, if it does not have any nodes,
it means that the current node used for the single NUMA node allocation,
otherwise it used for the multi-NUMA allocation.

## Method Changes
Now the `GetTopologyHints` method will return also multi-NUMA
hints, with some restrictions:

- preferred hints will be only single NUMA hints, with nodes that
not included under the multi-NUMA node `kubelet` flag and multi-NUMA
hints with nodes included under the flag
- if the node already has single NUMA node assignment, it can not be used for
multi-NUMA assignment and vice versa
- node can not be used for multi-NUMA assignment, if it already has assignment,
within different group

## Memory calculation

In cases when the multi-NUMA requested, the code will firstly check the first node, if it has enough memory, we will use the memory from it and update only one node information, otherwise it will allocate as much as possible updated the state of the node, and will move to the next node.

## TODO
- [x] adapt `Allocate` method
- [x] adapt `RemoveContainer` method

cynepco3hahue commented:
@cezaryzukowski @bg-chun The PR is ready for review, I lack unittests for new functionality, but I will create it under the separate PR, because this one already too big.

k-wiatrzyk commented on pkg/kubelet/cm/memorymanager/state/state.go:
```diff
@@ -29,16 +29,41 @@ type MemoryTable struct {
+	NumberOfAssignments int `json:"numberOfAssignments"`
+	// MemoryTable contains NUMA node memory related information
+	MemoryMap map[v1.ResourceName]*MemoryTable `json:"memoryMap"`
+	// NodeGroups contains NUMA nodes that current NUMA node in group with them
+	// It means that we have container that pinned to the current NUMA node and all group nodes
+	Nodes []int `json:"nodes"`
```
This is pretty hard to understand based on the description.
`Nodes` contains IDs of NUMA nodes that are in a group with the node that is described by `NodeState`. Is that right?

cynepco3hahue replied:
Yes, including that node that described by the `NodeState`
so for example if we have cross NUMA allocation from nodes `[0, 1]`, both nodes will have `[0, 1]` under the `Nodes` field

k-wiatrzyk replied:
Ok, got it.

k-wiatrzyk commented on pkg/kubelet/cm/memorymanager/policy_single_numa.go:
```diff
@@ -92,45 +103,81 @@ func (p *singleNUMAPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Co
+	// topology manager returns the hint that does not satisfy completely the container request
+	// we should extend this hint to the one who will satisfy the request and include the current hint
+	if !isAffinitySatisfyRequest(machineState, bestHint.NUMANodeAffinity, requestedResources) {
+		extendedHint, err := p.extendTopologyManagerHint(s, requestedResources, bestHint.NUMANodeAffinity)
+		if err != nil {
+			return err
 		}
 
-		size, succeed := q.AsInt64()
-		if !succeed {
-			return fmt.Errorf("[memorymanager] failed to represent quantity as int64")
+		if !extendedHint.Preferred && bestHint.Preferred {
+			return fmt.Errorf("[memorymanager] failed to find the extended preferred hint")
 		}
+		bestHint = extendedHint
+	}
```
As I understand extending hints is needed when **"restricted"** policy of Topology Manager is used. In case of other policies, i.e. **"single-numa-node"** this won't be used as TM will always return a hint that satisfies completely the container request or he rejects the pod admission. 

If that's right, I think it's hard to know the purpose of `extendTopologyManagerHint` only based on the code so there may be a good place for a broader explanation of that.

cynepco3hahue replied:
Added the proper description on top of `extendTopologyManagerHint`.

k-wiatrzyk commented on pkg/kubelet/cm/memorymanager/policy_single_numa.go:
```diff
@@ -40,12 +40,12 @@ type systemReservedMemory map[int]map[v1.ResourceName]uint64
 	// for example on the machine with four NUMA nodes and with this parameter equals to [[0,2]], only hints
 	// [0010, 1000, 0101] will be preferred
 	// if this parameter is empty, the default behavior will be used to set preferred hints, it means that only hint
 	// with a minimal amount of nodes that should satisfy the request will be set to preferred
```
+1 for description

I may be getting lost with that preferred. IIUC hint should be marked as `preferred` when it is the smallest possible quantity of nodes (based on amount of `allocatable` and not amount of free). 

So hint `[0101]` can be marked as `preferred` only if amount of memory in `allocatable` in each particular NUMA node is not enough for request. (.i.e. request>numa0.allocatable but request<numa0.allocatable+numa1.allocatable)

My suggestion:

> //multiNUMAGroups defines groups of NUMA nodes that will be treated as a joint pool of memory
> // for example on the machine with four NUMA nodes and with this parameter equals to [[0,2]], only hints
> // [0010, 1000, 0101] will be generated, where hints [0010,1000] are marked as preferred
> // if requested amount of memory can't be satisfied by theoretical capacity of each particular NUMA node but can be satisfied by multiNUMAGroup, then hint [0101] will be set as preferred
> // if this parameter is empty, each NUMA node is treated as separate entity

k-wiatrzyk commented on pkg/kubelet/cm/memorymanager/policy_single_numa.go:
```diff
@@ -622,7 +626,10 @@ func isAffinitySatisfyRequest(machineState state.NodeMap, mask bitmask.BitMask,
 }
 
-// hints for all memory types should be the same, so we will check hints only for regular memory type
+// extendTopologyManagerHint extends the topology manager hint, in case when it does not satisfy to the container request
+// the topology manager uses bitwise AND to merge all topology hints into the best one, so in case of the restricted policy,
+// it possible that we will get the subset of hint that we provided to the topology manager, in this case we want to extend
+// it to the original one
```
That is nice explanation

k-wiatrzyk commented on pkg/kubelet/cm/memorymanager/state/state.go:
```diff
@@ -36,8 +36,10 @@ type NodeState struct {
 	MemoryMap map[v1.ResourceName]*MemoryTable `json:"memoryMap"`
-	// NodeGroups contains NUMA nodes that current NUMA node in group with them
-	// It means that we have container that pinned to the current NUMA node and all group nodes
+	// Nodes contains the current NUMA node and all other nodes that current NUMA node in group with them
+	// This parameter used to indicate if the current node used for the multiple NUMA node memory allocation
+	// For example if some container has pinning 0,1,2, NUMA nodes 0,1,2 under the state will have
+	// this parameter equals to [0, 1, 2]
```
Very useful example. 

Small fixes:
> // Nodes contains the current NUMA node and all other nodes that are in a group with current NUMA node
> // This parameter indicates if the current node is used for the multiple NUMA node memory allocation

k-wiatrzyk commented on pkg/kubelet/cm/memorymanager/policy_single_numa.go:
```diff
@@ -357,44 +361,76 @@ func (p *singleNUMAPolicy) calculateHints(s state.State, requestedResources map[
 	}
 
-	return p.isMultipleNUMAGroup(maskBits)
+	return p.isMultiNUMAGroup(maskBits)
```
I think we should also check here if a group is a minimal possible hint to set it as preferred.
```
if len(maskBits) == minAffinitySize {...}
```

I understand it as following logic:
we have:

Case 1):
Memory allocatable: 0: [[10GB, 10GB, 10Gb, 10GB]
NUMA groups: [2,3]
Requirement: 5GB
Possible hints [0001-preferred,0010-preferred,1100]

Case 2):
Memory allocatable: 0: [[10GB, 10GB, 10Gb, 10GB]
NUMA groups: [2,3]
Requirement: 15GB
Possible hints [1100-preferred]

cynepco3hahue commented:
@k-wiatrzyk I Fixed small comments and dropped the multi NUMA node parameter, I will merge this one, just to finish all coding work in case if the KEP will get the approval.
